### ISSUE TITLE
segfault on :cscope show

### ISSUE THREAD

**Issue by mikewadsten** [neovim/neovim#4537]
- Neovim version: NVIM v0.1.3-387-g90e8bdc
- [x] Vim behaves differently?
    - Vim version: VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Mar  9 2016 16:33:05), Included patches: 1-1529
- Operating system/version: Arch Linux, 32-bit, kernel 4.4.5
- Terminal name/version: SSH, or Gnome Terminal

- `$TERM`: xterm-256color

### Actual behaviour

Upon running `:cs show`, Neovim segfaults, inside vim_vsnprintf.

### Expected behaviour

Output of `:cs show` appears in window, listing cscope connections.

### Steps to reproduce using `nvim -u NONE`

`nvim -u NONE`
```
:cs add /home/mike/work/redacted/.git/cscope /home/mike/work/redacted -ia
:cs show
```

(I use some Git hooks to run cscope over the repo, that's why the file is under .git)

This also happens when using `:set csprg=gtags-cscope` and `:cs add /home/mike/work/redacted/GTAGS /home/mike/work/redacted -ia`

### Captured GDB output

```
$ gdb /usr/local/bin/nvim
...
(gdb) r -u NONE
...
:cs add /home/mike/work/redacted/.git/cscope /home/mike/work/redacted -ia
:cs show
Thread 1 "nvim" received signal SIGSEGV, Segmentation fault.
0xb7ddcf36 in __strlen_sse2_bsf () from /usr/lib/libc.so.6
(gdb) bt
#0  0xb7ddcf36 in __strlen_sse2_bsf () from /usr/lib/libc.so.6
#1  0x0815bb0a in vim_vsnprintf (str=0x8349a60 <IObuff> " 0 -5234881856066808083  /home/mike/work/redacted", ' ' <repeats 12 times>, "\n\243pid\315\026\301\250encoding\304\005utf-8", str_m=1025, 
    fmt=0x82c1c20 "%2zu %-5lld  %-34s  %-32s", ap=0xbfffef38 "X\357\377\277\272\354\022\b\377\214#\b$g3\b3\242\202\267", tvs=0x0) at /home/mike/personal/neovim/src/nvim/message.c:3271
#2  0x08156bee in smsg (s=0x82c1c20 "%2zu %-5lld  %-34s  %-32s") at /home/mike/personal/neovim/src/nvim/message.c:328
#3  0x0812eda3 in cs_show (eap=0xbffff068) at /home/mike/personal/neovim/src/nvim/if_cscope.c:2085
#4  0x0812af98 in do_cscope_general (eap=0xbffff068, make_split=0) at /home/mike/personal/neovim/src/nvim/if_cscope.c:205
#5  0x0812afc4 in do_cscope (eap=0xbffff068) at /home/mike/personal/neovim/src/nvim/if_cscope.c:216
#6  0x080ea70a in do_one_cmd (cmdlinep=0xbffff12c, sourcing=0, cstack=0xbffff1f8, fgetline=0x80ff6d0 <getexline>, cookie=0x0) at /home/mike/personal/neovim/src/nvim/ex_docmd.c:2191
#7  0x080e7883 in do_cmdline (cmdline=0x0, fgetline=0x80ff6d0 <getexline>, cookie=0x0, flags=0) at /home/mike/personal/neovim/src/nvim/ex_docmd.c:601
#8  0x08177750 in nv_colon (cap=0xbffff648) at /home/mike/personal/neovim/src/nvim/normal.c:4485
#9  0x08170c52 in normal_execute (state=0xbffff5dc, key=58) at /home/mike/personal/neovim/src/nvim/normal.c:1147
#10 0x0821ba0a in state_enter (s=0xbffff5dc) at /home/mike/personal/neovim/src/nvim/state.c:55
#11 0x0816f26b in normal_enter (cmdwin=false, noexmode=false) at /home/mike/personal/neovim/src/nvim/normal.c:464
#12 0x08138c71 in main (argc=3, argv=0xbffff834) at /home/mike/personal/neovim/src/nvim/main.c:538
(gdb) up
#1  0x0815bb0a in vim_vsnprintf (str=0x8349a60 <IObuff> " 0 -5234881856066808083  /home/mike/work/redacted", ' ' <repeats 12 times>, "\n\243pid\315\026\301\250encoding\304\005utf-8", str_m=1025, 
    fmt=0x82c1c20 "%2zu %-5lld  %-34s  %-32s", ap=0xbfffef38 "X\357\377\277\272\354\022\b\377\214#\b$g3\b3\242\202\267", tvs=0x0) at /home/mike/personal/neovim/src/nvim/message.c:3271
3271	            str_arg_l = strlen(str_arg);
(gdb) p str_arg
$1 = 0x20 <error: Cannot access memory at address 0x20>
(gdb) info args
str = 0x8349a60 <IObuff> " 0 -5234881856066808083  /home/mike/work/redacted", ' ' <repeats 12 times>, "\n\243pid\315\026\301\250encoding\304\005utf-8"
str_m = 1025
fmt = 0x82c1c20 "%2zu %-5lld  %-34s  %-32s"
ap = 0xbfffef38 "X\357\377\277\272\354\022\b\377\214#\b$g3\b3\242\202\267"
tvs = 0x0
(gdb) info locals
space_for_positive = 1
str_arg_l = 1
precision = 0
alternate_form = 0
force_sign = 0
uchar_arg = 8 '\b'
number_of_zeros_to_pad = 0
fmt_spec = 115 's'
zero_padding = 0
precision_specified = 0
justify_left = 1
tmp = "-5234618591751418664\000\000(\355\377\277\002\000\000\000\001\000\000\000ԏ#\b\300\021\203\267\060\242\202\267\000\371`\266\001\000\000\000p\355\377\277x\355\377\277\270\217#\b\001\000\000\000ԏ#\b\300\021\203\267\001\000\000\000\035;,\b\001\000\000\000 \000\000\000\000\000\000\000\060\242\202\267\001", '\000' <repeats 11 times>, "\001\000\000\000\001\000\000\000H\356\377\277\300\021\203\267h\000\000\000\000\000\000\000\036;,\b \000\000\000\000\200\361\267\000\200\361\267\210\355\377\277\310/\025\b\031\000\000\000p\355\377\277\230\355\377\277\064z\f\b\004W3\b`K4\b\200\315Z\267\031\000\000\000\300\021\203\267\001\000\000\000"...
zero_padding_insertion_ind = 0
min_field_width = 32
length_modifier = 0 '\000'
str_arg = 0x20 <error: Cannot access memory at address 0x20>
str_l = 61
str_avail = true
p = 0x82c3b38 "s"
arg_idx = 1
__PRETTY_FUNCTION__ = "vim_vsnprintf"
```

Interestingly, on another machine, I generally see `str_arg` as `0xb0`.

**Comment by oni-link** [comment]
Cannot reproduce the problem.  

From the backtrace it looks like `csinfo[0].ppath` is somewhere set to the invalid address `0x20` (`0xb0`). Perhaps it would help to use `watch csinfo[0].ppath` in gdb and see where the invalid address is set.

Is the flag `-ia` necessary to trigger the problem?

**Comment by mikewadsten** [comment]
The flags are not necessary to reproduce - I get a segfault if I leave the flags out. The backtrace at that point is identical, and `str_arg` is `0x20`.

If I leave out the pre-path argument (the directory), I don't get a segfault but I do see some corrupted output:

```
 # pid    database name                       prepend path
 0 -5269273823869380275  U<89><e5><83><ec>^X<8b>E^H<8b>                          <none>
```

If I `b cs_show` when I reproduce, I can see that `csinfo[0].ppath` has the correct value going into the `smsg` call. I'll try some different watches and breakpoints in `vim_vsnprintf` to try to determine where things get messed up.

**Comment by mikewadsten** [comment]
I may have found the problem.

`src/nvim/if_cscope.c` starting at line 2084 does this:

```c
      if (csinfo[i].ppath != NULL)
        (void)smsg("%2zu %-5" PRId64 "  %-34s  %-32s",
            i, (long)csinfo[i].pid, csinfo[i].fname, csinfo[i].ppath);
      else
        (void)smsg("%2zu %-5" PRId64 "  %-34s  <none>",
            i, (long)csinfo[i].pid, csinfo[i].fname);
```

In the case I documented above, ppath is null, so we execute the else clause. Note that it's casting `csinfo[0].pid` to a long and passing it in for a `PRId64` argument (which on my system, 32-bit Linux with GCC 5.3.0, is `lld`).

Down in `vim_vsnprintf`, the format `%-5lld` sets up the following local variables:

Variable name | Value
---------- | ---
`fmt_spec` | `'d'`
`length_modifier` | `2`

...which means we run this case, `src/nvim/message.c` line 3361:

```c
          case '2':
            long_long_arg = tvs ? tv_nr(tvs, &arg_idx)
                                : va_arg(ap, long long int); // NOLINT (runtime/int)
            if (long_long_arg > 0) {
              arg_sign =  1;
            } else if (long_long_arg < 0) {
              arg_sign = -1;
            }
            break;
```

Note the `va_arg(ap, long long int)` call.

Upon first entering `vim_vsnprintf`, I would run `x/4x ap` and see some output like this:

```
0xbfffee9a:     0x00000000     0x00006157     0xb75ea100     0x08238cff
```

The first byte (0) is the value of `i`, the second value (0x6157 or 24919) is in fact the PID of the child cscope process, and the third value (0xb75ea100) is `csinfo[i].fname`.

`sizeof(long) = 4` and `sizeof(long long int) = 8` on my system, so `vim_vsnprintf` is fetching 64 bits of data where we passed in 32. As a result, `long_long_arg` has the value `0xb75ea10000006157`, or -5233568695585840809.

In conclusion, it's a type issue.

**Comment by oni-link** [comment]
@mikewadsten, here is a patch for what you found. Does this fix the problem on 32bit?

```diff
diff --git a/src/nvim/if_cscope.c b/src/nvim/if_cscope.c
index 45764d3..e82f3ef 100644
--- a/src/nvim/if_cscope.c
+++ b/src/nvim/if_cscope.c
@@ -2064,12 +2064,13 @@ static int cs_show(exarg_T *eap)
       if (csinfo[i].fname == NULL)
         continue;
 
-      if (csinfo[i].ppath != NULL)
-        (void)smsg("%2zu %-5" PRId64 "  %-34s  %-32s",
-            i, (long)csinfo[i].pid, csinfo[i].fname, csinfo[i].ppath);
-      else
-        (void)smsg("%2zu %-5" PRId64 "  %-34s  <none>",
-            i, (long)csinfo[i].pid, csinfo[i].fname);
+      if (csinfo[i].ppath != NULL) {
+        (void)smsg("%2zu %-5" PRId64 "  %-34s  %-32s", i,
+                   (int64_t)csinfo[i].pid, csinfo[i].fname, csinfo[i].ppath);
+      } else {
+        (void)smsg("%2zu %-5" PRId64 "  %-34s  <none>", i,
+                   (int64_t)csinfo[i].pid, csinfo[i].fname);
+      }
     }
   }

```

**Comment by mikewadsten** [comment]
@oni-link Yes, that works for me.

Should I submit the patch or do you want to?

**Comment by oni-link** [comment]
Please submit it. After all you have done all the work.
